File: BOJ/3987.cpp
/*

시작 위치에서 4방향 모두 시그널을 보내보고 전파 시간이 가장 긴 놈을 출력.

*/
#include<bits/stdc++.h>
using namespace std;

int N, M;
vector<string> Map;
vector<pair<char, int> > answer; // (시작방향, 전파시간) 쌍을 저장하기 위한 vector

// 전파 시간이 긴 순으로 내림차순 정렬
bool comp(pair<char, int> a, pair<char, int> b) {
	return a.second > b.second;
}

int signal(int y, int x, char direction) {
	int prop = 0;
	while (true) {
		// 탈출 조건
		if (prop > N*M) break; // 항성계의 전체 크기보다 크다면 무한루프를 돌고 있는 것으로 판단.
		if (x < 0 || y < 0 || x >= M || y >= N || Map[y][x] == 'C') // 맵 밖으로 나가거나 블랙홀을 만난 경우
			break;

		prop++;

		// 진행 방향에 따라 1초씩 이동한다.
		if (direction == 'U') {
			if (Map[y][x] == '/') { // 진행 방향이 U이고 '/' 행성을 만났을 때, 다음 진행 방향은 R
				direction = 'R';
				x += 1;
			}
			else if (Map[y][x] == '\\') { // 진행 방향이 U이고 '\' 행성을 만났을 때, 다음 진행 방향은 L
				direction = 'L';
				x -= 1;
			}
			else y -= 1; // 행성이 없을 경우 원래 진행 방향.
		}
		else if (direction == 'R') {
			if (Map[y][x] == '/') {
				direction = 'U';
				y -= 1;
			}
			else if (Map[y][x] == '\\') {
				direction = 'D';
				y += 1;
			}
			else x += 1;
		}
		else if (direction == 'D') {
			if (Map[y][x] == '/') {
				direction = 'L';
				x -= 1;
			}
			else if (Map[y][x] == '\\') {
				direction = 'R';
				x += 1;
			}
			else y += 1;
		}
		else {
			if (Map[y][x] == '/') {
				direction = 'D';
				y += 1;
			}
			else if (Map[y][x] == '\\') {
				direction = 'U';
				y -= 1;
			}
			else x -= 1;
		}
	}

	return prop;
}

int main() {
	cin >> N >> M;
	for (int i = 0; i < N; i++) {
		string s;
		cin >> s;
		Map.push_back(s);
	}

	int sy, sx;
	cin >> sy >> sx;
	sy--; // 문제에서 index가 1부터 시작하기 때문에 -1해줌.
	sx--;

	// 시작 지점에서 4방향으로 진행했을 때의 거리를 저장해준다.
	answer.push_back(make_pair('U', signal(sy, sx, 'U')));
	answer.push_back(make_pair('R', signal(sy, sx, 'R')));
	answer.push_back(make_pair('D', signal(sy, sx, 'D')));
	answer.push_back(make_pair('L', signal(sy, sx, 'L')));

	sort(answer.begin(), answer.end(), comp);

	cout << answer[0].first << '\n';
	if (answer[0].second > N*M) cout << "Voyager" << '\n';
	else cout << answer[0].second << '\n';

	return 0;
}

File: BOJ/python/3987.py
"""

재채점 되어서 틀렸던 문제.
line 18에서 항성계 전체를 도는 시간으로 탈출 조건을 걸어 두었는데
같은 지점을 방문하더라도 전파 방향이 다르다면 다른 경우인 것을 인지하지 못해서 틀림.
모든 방향을 고려하기 위해 간단히 *4를 해주었다.

"""
import sys

input = sys.stdin.readline


def signal(y, x, direction):
    prop = 0
    while True:
        global N, M
        if prop > N * M * 4:  # 전파 시간이 항성계 전체를 도는 시간(*4는 모든 방향)보다 크다면 무한루프라고 판단.
            break
        if (x < 0) or (y < 0) or (x >= M) or (y >= N) or (Map[y][x] == "C"):
            break  # 전파가 맵 밖으로 나가거나 블랙홀을 만난 경우
        prop += 1

        # 진행 방향에 따라 1초씩 이동한다.
        if direction == "U":
            if Map[y][x] == "/":
                direction = "R"
                x += 1
            elif Map[y][x] == "\\":
                direction = "L"
                x -= 1
            else:
                y -= 1
        elif direction == "R":
            if Map[y][x] == "/":
                direction = "U"
                y -= 1
            elif Map[y][x] == "\\":
                direction = "D"
                y += 1
            else:
                x += 1
        elif direction == "D":
            if Map[y][x] == "/":
                direction = "L"
                x -= 1
            elif Map[y][x] == "\\":
                direction = "R"
                x += 1
            else:
                y += 1
        else:
            if Map[y][x] == "/":
                direction = "D"
                y += 1
            elif Map[y][x] == "\\":
                direction = "U"
                y -= 1
            else:
                x -= 1
    return prop


global N, M, Map
N, M = map(int, input().rstrip("\n").split(" "))
Map = [input().rstrip("\n") for _ in range(N)]
PR, PC = map(int, input().rstrip("\n").split(" "))
PR -= 1 # index가 1부터 시작하기 때문에 -1을 해주었다.
PC -= 1

answer = dict()

# 시작 지점에서 4방향으로 모두 전파해보고 거리를 dict 형태로 저장해준다.
answer["U"] = signal(PR, PC, "U")
answer["R"] = signal(PR, PC, "R")
answer["D"] = signal(PR, PC, "D")
answer["L"] = signal(PR, PC, "L")

dict_key_max = max(answer.keys(), key=lambda x: answer[x]) # dict에서 가장 큰 value값을 가지는 key를 찾는다.
print(dict_key_max)
if answer[dict_key_max] > N * M * 4:
    print("Voyager")
else:
    print(answer[dict_key_max])
